### ISSUE TITLE
Incorrect error on dynamic class property name

### ISSUE THREAD

**Issue by rotu** [microsoft/TypeScript#56523]
### 🔎 Search Terms

computed property name, TS1166, 1166

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

https://www.typescriptlang.org/play?target=99&ssl=8&ssc=1&pln=1&pc=1&filetype=js#code/DYUwLgBAHhC8ECIBmB7FCBQBjFA7AzpALYCeA8gEYBWIWk8A3hhBANpQC6AXBAIwYBfbMACG+fBADKKIiADCo8RCYt2HOH0EYgA

### 💻 Code

```ts repro
let x = "foo"

const someObject = {
  [x]: 1
}

class SomeClass {
  [x] = 1
}

```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?ts=5.3.2#)

### 🙁 Actual behavior

Typescript emits a TS1166 error on the class assignment which makes it seem like this code is invalid:

> A computed property name in a class property declaration must have a simple literal type or a 'unique symbol' type.

Note that the object assignment is correctly handled. The class definition results in an error.

### 🙂 Expected behavior

I expect Typescript to accept the syntactically valid class definition, similar to how it accepts the object literal.

**Comment by MartinJohns** [comment]
> I expect Typescript to allow this code, as it is valid JavaScript.

With that argument you can just use JavaScript directly instead of TypeScript, because code with type errors is also just "valid JavaScript". See also: https://stackoverflow.com/a/4175039

**Comment by fatcerberus** [comment]
@MartinJohns I think you linked the wrong SO post... the one you posted is something about SQL.

**Comment by rotu** [comment]
One of TypeScript's goals is to "Align with current and future ECMAScript proposals." I suspect the current behavior is because TypeScript wants very badly for a "class" to be statically analyzable although a class can indeed be changed.

The error incorrectly states that the code is invalid ("must"). At the minimum, it should be softened to something like "cannot infer the type of a dynamically defined property name". But I suspect the current behavior is a carryover from before `class` was as mature.

For contrast, I could have written `SomeClass.prototype.x = 1` and TypeScript would have accepted it as okay.

**Comment by rotu** [comment]
Related: #27864

**Comment by Josh-Cena** [comment]
What is your intended behavior for its types? An object with such a computed property is simply demoted to a `Record<string, number>`, so a class would also become `new () => Record<string, number>`, which means you can't have any methods on the class whatsoever, and that doesn't seem hugely useful. `SomeClass.prototype = { [x]: 1 }` is fine simply because TS doesn't have the concept of prototypes and just treats `prototype` as a normal data property.

**Comment by rotu** [comment]
@Josh-Cena yes, I don’t expect Typescript to meaningfully infer anything here besides that `new SomeClass` is an `object`. I was trying to do a sort of class mix-in here, and gradually type things later.

**Comment by RyanCavanaugh** [comment]
> The error incorrectly states that the code is invalid ("must")

This is how nearly every type-related error is phrased. I don't think your expectations are properly aligned here.

**Comment by rotu** [comment]
I think the confusion comes from the fact that a `class` has particular JavaScript runtime semantics. Typescript deriving from that a `type` which has stricter semantics based on its static analysis. But then it's making those seem like invalid use of the underlying JavaScript (instead of TypeScript's imputed types).

This is true *even when*:

- flow-control analysis can narrow the property to a string literal (as above since `x` is captured by value, not by reference), or:
  ```ts
  if (x === 'foo'){
    class B {
      [x] = 1
    }
    console.log(new B())
  }
  ```
- generic instantiation can narrow the property to a string literal:
  ```ts
  const foo = <const X extends string,>(x:X) => (class{[x]=1})
  ```
- the property is declared `static`.
  ```ts
  const MyNamespace = class
  ```
